qemu/qemu.go

```diff
@@ -26,14 +26,876 @@ package qemu
+	params = append(params, strings.Join(qmpParams, ""))
+
+	return params
+}
+
+func appendObject(params []string, object Object) ([]string, error) {
+	if object.Valid() == false {
+		return nil, fmt.Errorf("Invalid Object")
+	}
+
+	var objectParams []string
+
+	objectParams = append(objectParams, string(object.Type))
+
+	switch object.Type {
+	case MemoryBackendFile:
+		objectParams = append(objectParams, fmt.Sprintf(",id=%s", object.ID))
+		objectParams = append(objectParams, fmt.Sprintf(",mem-path=%s", object.MemPath))
+		objectParams = append(objectParams, fmt.Sprintf(",size=%d", object.Size))
+	}
+
+	params = append(params, "-object")
+	params = append(params, strings.Join(objectParams, ""))
+
+	return params, nil
+}
+
+func appendFilesystemDevice(params []string, fsdev FSDevice) ([]string, error) {
+	if fsdev.Valid() == false {
+		return nil, fmt.Errorf("Invalid filesystem device")
+	}
+
+	var fsParams []string
+
+	fsParams = append(fsParams, string(fsdev.Driver))
+	fsParams = append(fsParams, fmt.Sprintf(",id=%s", fsdev.ID))
+	fsParams = append(fsParams, fmt.Sprintf(",path=%s", fsdev.Path))
+	fsParams = append(fsParams, fmt.Sprintf(",security-model=%s", fsdev.SecurityModel))
+
+	params = append(params, "-fsdev")
+	params = append(params, strings.Join(fsParams, ""))
+
+	return params, nil
+}
+
+func appendCharDevice(params []string, cdev CharDevice) ([]string, error) {
+	if cdev.Valid() == false {
+		return nil, fmt.Errorf("Invalid character device")
+	}
+
+	var cdevParams []string
+
+	cdevParams = append(cdevParams, string(cdev.Backend))
+	cdevParams = append(cdevParams, fmt.Sprintf(",id=%s", cdev.ID))
+	cdevParams = append(cdevParams, fmt.Sprintf(",path=%s", cdev.Path))
+
+	params = append(params, "-chardev")
+	params = append(params, strings.Join(cdevParams, ""))
+
+	return params, nil
+}
+
+func appendNetDevice(params []string, netdev NetDevice) ([]string, error) {
+	if netdev.Valid() == false {
+		return nil, fmt.Errorf("Invalid network device")
+	}
+
+	var netdevParams []string
+
+	netdevParams = append(netdevParams, string(netdev.Type))
+	netdevParams = append(netdevParams, fmt.Sprintf(",id=%s", netdev.ID))
+	netdevParams = append(netdevParams, fmt.Sprintf(",ifname=%s", netdev.IFName))
+
+	if netdev.DownScript != "" {
+		netdevParams = append(netdevParams, fmt.Sprintf(",downscript=%s", netdev.DownScript))
+	}
+
+	if netdev.Script != "" {
+		netdevParams = append(netdevParams, fmt.Sprintf(",script=%s", netdev.Script))
+	}
+
+	if len(netdev.FDs) > 0 {
+		var fdParams []string
+
+		for _, fd := range netdev.FDs {
+			fdParams = append(fdParams, fmt.Sprintf("%d", fd))
+		}
+
+		netdevParams = append(netdevParams, fmt.Sprintf(",fds=%s", strings.Join(fdParams, ":")))
+	}
+
+	if netdev.VHost == true {
+		netdevParams = append(netdevParams, ",vhost=on")
+	}
+
+	params = append(params, "-netdev")
+	params = append(params, strings.Join(netdevParams, ""))
+
+	return params, nil
+}
+
+func appendDrivers(params []string, config Config) []string {
```

Comment:
@sameo may make sense to also have a kernel append function also for each device. hence they can have two append interfaces. qemuAppend and kernelAppend.